### PR TITLE
Extract memory manager into a separate module

### PR DESCRIPTION
**Pull request by projedi**

### CODE REVIEW

File: shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/KonanTargetExtenstions.kt [hunk 1]
```diff
@@ -20,4 +22,10 @@ fun KonanTarget.supportsMimallocAllocator(): Boolean =
         is KonanTarget.IOS_ARM64 -> true
         is KonanTarget.IOS_X64 -> true
         else -> false // watchOS/tvOS/android_x86/android_arm32 aren't tested; linux_mips32/linux_mipsel32 need linking with libatomic.
-    }
\ No newline at end of file
+    }
+
+fun KonanTarget.supportsThreads(): Boolean =
+     when(this) {
+        is KonanTarget.WASM32 -> false
```
Comment (sbogolepov) [inline shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/KonanTargetExtenstions.kt:29]: Zephyr as well.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/KonanConfig.kt [hunk 2]
```diff
@@ -115,17 +115,51 @@ class KonanConfig(val project: Project, val configuration: CompilerConfiguration
 
     internal val runtimeNativeLibraries: List<String> = mutableListOf<String>().apply {
         add(if (debug) "debug.bc" else "release.bc")
-        add(if (memoryModel == MemoryModel.STRICT) "strict.bc" else "relaxed.bc")
-        if (shouldCoverLibraries || shouldCoverSources) add("profileRuntime.bc")
-        if (configuration.get(KonanConfigKeys.ALLOCATION_MODE) == "mimalloc") {
-            if (!target.supportsMimallocAllocator()) {
+        val useMimalloc = if (configuration.get(KonanConfigKeys.ALLOCATION_MODE) == "mimalloc") {
+            if (target.supportsMimallocAllocator()) {
                 configuration.report(CompilerMessageSeverity.STRONG_WARNING,
```
[the remark this answers is not in the document]
Comment (projedi) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/KonanConfig.kt]: Right, thanks!

File: shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/KonanTargetExtenstions.kt (hunk 1 above)
Comment (projedi) [reply shared/src/main/kotlin/org/jetbrains/kotlin/konan/target/KonanTargetExtenstions.kt:29]: Missed it, thanks!

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/KonanConfig.kt [hunk 3]
```diff
@@ -115,17 +115,51 @@ class KonanConfig(val project: Project, val configuration: CompilerConfiguration
...
+            false
+        }
+        val effectiveMemoryModel = when (memoryModel) {
+            MemoryModel.STRICT -> MemoryModel.STRICT
+            MemoryModel.RELAXED -> MemoryModel.RELAXED
+            MemoryModel.EXPERIMENTAL -> {
+                if (!useMimalloc) {
+                    configuration.report(CompilerMessageSeverity.STRONG_WARNING,
+                            "Experimental memory model requires mimalloc allocator. Used strict memory model.")
+                    MemoryModel.STRICT
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/KonanConfig.kt:136]: This behaviour unnecessarily complicates enabling experimental. Why not just use mimalloc if experimental "memory model" is selected? Is this actually user visible?

**Comment by SvyatoslavScherbina** [comment]
It makes some sense to make runtime directory names shorter. Something like `legacymm/` and `newmm/`?

File: runtime/src/legacy_memory_manager/cpp/Memory.cpp [hunk 4]
```diff
@@ -430,8 +437,70 @@ inline bool isShareable(ContainerHeader* container) {
     return container == nullptr || container->shareable();
 }
 
+MetaObjHeader* createMetaObject(TypeInfo** location);
```
Comment (SvyatoslavScherbina) [inline runtime/src/legacy_memory_manager/cpp/Memory.cpp:440]: Meta objects are likely required for new MM too.

File: runtime/src/legacy_memory_manager/cpp/Memory.cpp [hunk 5]
```diff
@@ -430,8 +437,70 @@ inline bool isShareable(ContainerHeader* container) {
...
+}
+
+ALWAYS_INLINE bool isPermanentOrFrozen(const ObjHeader* obj) {
+    auto* container = containerFor(obj);
+    return container == nullptr || container->frozen();
+}
+
+ALWAYS_INLINE bool isShareable(const ObjHeader* obj) {
+    auto* container = containerFor(obj);
+    return container == nullptr || container->shareable();
```
Comment (SvyatoslavScherbina) [inline runtime/src/legacy_memory_manager/cpp/Memory.cpp:478]: AFAIK, none of the usages needs this `nullptr` check.

File: runtime/src/legacy_memory_manager/cpp/Memory.cpp [hunk 6]
```diff
@@ -46,6 +46,11 @@
 #include "Runtime.h"
 #include "Utils.h"
 #include "WorkerBoundReference.h"
+#include "Weak.h"
+
+#ifdef KONAN_OBJC_INTEROP
+#include "ObjCExport.h"
```
Comment (SvyatoslavScherbina) [inline runtime/src/legacy_memory_manager/cpp/Memory.cpp:52]: Somewhat innapropriate here.

File: runtime/src/legacy_memory_manager/cpp/Memory.cpp (hunk 5 above)
Comment (SvyatoslavScherbina) [reply runtime/src/legacy_memory_manager/cpp/Memory.cpp:478]: Well, almost none.

File: runtime/src/main/cpp/Arrays.cpp [hunk 7]
```diff
@@ -29,8 +29,8 @@ namespace {
 
 ALWAYS_INLINE inline void mutabilityCheck(KConstRef thiz) {
   // TODO: optimize it!
-  if (!thiz->local() && thiz->container()->frozen()) {
-    ThrowInvalidMutabilityException(thiz);
+  if (!thiz->local() && isPermanentOrFrozen(thiz)) {
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/Arrays.cpp:32]: No null check required for container here.

File: runtime/src/legacy_memory_manager/cpp/Memory.cpp (hunk 4 above)
Comment (SvyatoslavScherbina) [reply runtime/src/legacy_memory_manager/cpp/Memory.cpp:440]: Also some of the meta objects machinery is still in common `Memory.h`.
Does anything prevent us from sharing the entire implementation for now?

File: runtime/src/main/cpp/ObjCExport.h [hunk 8]
```diff
@@ -3,6 +3,8 @@
 
 #if KONAN_OBJC_INTEROP
 
+#if __OBJC__
```
Comment (SvyatoslavScherbina) [inline runtime/src/main/cpp/ObjCExport.h:6]: Why is this required?

File: runtime/src/legacy_memory_manager/cpp/Memory.cpp (hunk 6 above)
Comment (SvyatoslavScherbina) [reply runtime/src/legacy_memory_manager/cpp/Memory.cpp:52]: Maybe make a smaller header dedicated to MM-ObjC integration?

File: runtime/src/relaxed/cpp/MemoryImpl.cpp [hunk 9]
```diff
@@ -3,7 +3,7 @@
  * that can be found in the LICENSE file.
  */
 #include "Memory.h"
-#include "MemoryPrivate.hpp"
+#include "../../legacy_memory_manager/cpp/MemoryPrivate.hpp" // Fine, because this module is a part of legacy MM.
```
Comment (SvyatoslavScherbina) [inline runtime/src/relaxed/cpp/MemoryImpl.cpp:6]: Can we move `strict` and `relaxed` into `legacy_memory_manager` later?

Comment (projedi) [reply runtime/src/relaxed/cpp/MemoryImpl.cpp:6]: Yes, we can make `legacy_memory_manager` depend upon a flag `USE_STRICT_MEMORY_MODEL` and compile it twice with 2 different values.

File: runtime/src/main/cpp/ObjCExport.h (hunk 8 above)
Comment (projedi) [reply runtime/src/main/cpp/ObjCExport.h:6]: To be includable into `Memory.cpp`.

File: runtime/src/relaxed/cpp/MemoryImpl.cpp (hunk 9 above)
Comment (SvyatoslavScherbina) [reply runtime/src/relaxed/cpp/MemoryImpl.cpp:6]: I actually meant just moving the directories, not merging the modules.

File: runtime/src/main/cpp/ObjCExport.h (hunk 8 above)
Comment (SvyatoslavScherbina) [reply runtime/src/main/cpp/ObjCExport.h:6]: Well, this gives one more reason for https://github.com/JetBrains/kotlin-native/pull/4446#discussion_r511776827.

File: runtime/src/relaxed/cpp/MemoryImpl.cpp (hunk 9 above)
Comment (projedi) [reply runtime/src/relaxed/cpp/MemoryImpl.cpp:6]: We will need to redo our c++ build plugin for that, but it's possible.

File: runtime/src/main/cpp/ObjCExport.h (hunk 8 above)
Comment (projedi) [reply runtime/src/main/cpp/ObjCExport.h]: Done

File: runtime/src/main/cpp/Arrays.cpp (hunk 7 above)
Comment (projedi) [reply runtime/src/main/cpp/Arrays.cpp]: Right, removed null check

File: runtime/src/legacy_memory_manager/cpp/Memory.cpp (hunk 6 above)
Comment (projedi) [reply runtime/src/legacy_memory_manager/cpp/Memory.cpp]: Yeah, good idea, thanks!

File: runtime/src/legacy_memory_manager/cpp/Memory.cpp (hunk 4 above)
Comment (projedi) [reply runtime/src/legacy_memory_manager/cpp/Memory.cpp]: Right, shared some of it.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/KonanConfig.kt (hunk 3 above)
Comment (projedi) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/KonanConfig.kt:125]: Good idea, reworked.

File: runtime/src/legacy_memory_manager/cpp/Memory.cpp (hunk 5 above)
Comment (projedi) [reply runtime/src/legacy_memory_manager/cpp/Memory.cpp]: Removed here too.

**Comment by projedi** [comment]
> It makes some sense to make runtime directory names shorter. Something like `legacymm/` and `newmm/`?

Renamed to `legacymm/` and `mm/`

**Review by SvyatoslavScherbina — APPROVED** [review]
